cmd/cloudwatcher/metrics.go

```diff
@@ -0,0 +1,506 @@
+			&cloudwatch.MetricDatum{
+				Timestamp:         aws.Time(timestamp),
+				MetricName:        aws.String(fmt.Sprintf("JVM%sPoolPeakUsed", name)),
+				Dimensions:        dimensions,
+				Unit:              aws.String(cloudwatch.StandardUnitBytes),
+				StorageResolution: aws.Int64(1),
+				Value:             aws.Float64(peakUsed),
+			},
+			&cloudwatch.MetricDatum{
+				Timestamp:         aws.Time(timestamp),
+				MetricName:        aws.String(fmt.Sprintf("JVM%sPoolUtilization", name)),
+				Dimensions:        dimensions,
+				Unit:              aws.String(cloudwatch.StandardUnitPercent),
+				StorageResolution: aws.Int64(1),
+				Value:             aws.Float64(used / maxJVMHeapBytes * 100), // CloudWatch percents are int 0-100
+			},
+		)
+	}
+
+	// Sum garbage collection stats.
+	for name, c := range s.garbageCollectors {
+		count := sum(c.Count...)
+		time := sum(c.Time...)
+
+		name = strings.Title(name)
+		metrics = append(metrics,
+			&cloudwatch.MetricDatum{
+				Timestamp:         aws.Time(timestamp),
+				MetricName:        aws.String(fmt.Sprintf("GC%sCount", name)),
+				Dimensions:        dimensions,
+				Unit:              aws.String(cloudwatch.StandardUnitCount),
+				StorageResolution: aws.Int64(1),
+				Value:             aws.Float64(count),
+			},
+			&cloudwatch.MetricDatum{
+				Timestamp:         aws.Time(timestamp),
+				MetricName:        aws.String(fmt.Sprintf("GC%sTime", name)),
+				Dimensions:        dimensions,
+				Unit:              aws.String(cloudwatch.StandardUnitCountSecond),
+				StorageResolution: aws.Int64(1),
+				Value:             aws.Float64(time / 1000),
+			},
+		)
+	}
+
+	// Sum filesystem utilization stats.
+	if anyNonZero(s.maxFSBytes...) {
+		sumMaxFSBytes := sum(s.maxFSBytes...)
+		sumAvailableFSBytes := sum(s.availableFSBytes...)
+
+		metrics = append(metrics,
+			&cloudwatch.MetricDatum{
+				Timestamp:         aws.Time(timestamp),
+				MetricName:        aws.String("FSMaxBytes"),
+				Dimensions:        dimensions,
+				Unit:              aws.String(cloudwatch.StandardUnitBytes),
+				StorageResolution: aws.Int64(1),
+				Value:             aws.Float64(sumMaxFSBytes),
+			},
+			&cloudwatch.MetricDatum{
+				Timestamp:         aws.Time(timestamp),
+				MetricName:        aws.String("FSAvailableBytes"),
+				Dimensions:        dimensions,
+				Unit:              aws.String(cloudwatch.StandardUnitBytes),
+				StorageResolution: aws.Int64(1),
+				Value:             aws.Float64(sumAvailableFSBytes),
+			},
+			&cloudwatch.MetricDatum{
+				Timestamp:         aws.Time(timestamp),
+				MetricName:        aws.String("FSUtilization"),
+				Dimensions:        dimensions,
+				Unit:              aws.String(cloudwatch.StandardUnitPercent),
+				StorageResolution: aws.Int64(1),
+				Value:             aws.Float64((1.0 - (sumAvailableFSBytes / sumMaxFSBytes)) * 100), // CloudWatch percents are int 0-100
+			},
+		)
+	}
+
+	return metrics
+}
+
+// sum returns a sum of d.
+func sum(d ...float64) float64 {
```

Comment:
Why do you use variatic args instead of passing a slice? Is there an advantage either way?

Reply:
Variadic lets you use it for both slices (`sum(d...)`) and a static number of variables (`sum(a, b, c)`). It's all slices behind the scenes anyway.

Reply:
Would it be fair to say using variadic here is a personal preference? Just curious.

Reply:
Yeah.